contracts/nexus-gateway/src/nexus.rs

```diff
@@ -0,0 +1,74 @@
+use axelar_wasm_std::nonempty;
+use connection_router::state::{
+    Address, ChainName, Message as RouterMessage, MessageId, ID_SEPARATOR,
+};
+use cosmwasm_std::{CosmosMsg, CustomMsg};
+use error_stack::{Report, Result, ResultExt};
+use hex::FromHex;
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+use crate::error::ContractError;
+
+const ZEROX_PREFIX: &str = "0x";
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+// this matches the message type defined in the nexus module
+// https://github.com/axelarnetwork/axelar-core/blob/6c887df3797ba660093061662aff04e325b9c429/x/nexus/exported/types.pb.go#L405
+pub struct Message {
+    pub source_chain: ChainName,
+    pub source_address: Address,
+    pub destination_chain: ChainName,
+    pub destination_address: Address,
+    pub payload_hash: nonempty::Vec<u8>,
+    pub source_tx_id: nonempty::Vec<u8>,
+    pub source_tx_index: u64,
+}
+
+impl CustomMsg for Message {}
+
+fn parse_message_id(message_id: &MessageId) -> Result<(nonempty::Vec<u8>, u64), ContractError> {
+    // expected format: <tx_id>:<index>
+    let components = message_id.split(ID_SEPARATOR).collect::<Vec<_>>();
+
+    if components.len() != 2 {
+        return Err(ContractError::InvalidMessageId(message_id.clone()).into());
+    }
+
+    // TODO: decode differently depending on the chain?
+    let tx_id = <Vec<u8>>::from_hex(components[0].trim_start_matches(ZEROX_PREFIX))
+        .change_context_lazy(|| ContractError::InvalidMessageId(message_id.clone()))?;
+    let tx_id: nonempty::Vec<u8> = <nonempty::Vec<u8>>::try_from(tx_id)
+        .change_context_lazy(|| ContractError::InvalidMessageId(message_id.clone()))?;
+    let index = components[1]
+        .parse::<u64>()
+        .change_context_lazy(|| ContractError::InvalidMessageId(message_id.clone()))?;
+
+    Ok((tx_id, index))
+}
+
+impl TryFrom<RouterMessage> for Message {
+    type Error = Report<ContractError>;
+
+    fn try_from(msg: RouterMessage) -> Result<Self, ContractError> {
+        let (source_tx_id, source_tx_index) = parse_message_id(&msg.cc_id.id)?;
```

Comment:
Isn't there supposed to be some sort of fallback if the message id can't be parsed into this format? The source chain can use whatever message id format it wants, it might not fit into this format.